File: packages/dart/lib/src/mailbox.dart
import 'event_type.dart';
import 'event.dart';
import 'rule.dart';

/// Mailbox -- holds pending events for a given [EventType].
/// Events remain in the mailbox until all registered consumers have processed them.
class Mailbox<T> {
  final EventType<T> type;
  final List<PulseEvent<T>> queue = [];

  Mailbox(this.type);

  /// Add an event to this mailbox.
  void enqueue(PulseEvent<T> event) {
    queue.add(event);
  }

  /// Check if there is at least one event ready for the given rule.
  bool hasReadyEvent(Rule rule) {
    for (var i = 0; i < queue.length; i++) {
      if (queue[i].pendingConsumers.contains(rule)) {
        return true;
      }
    }
    return false;
  }

  /// Count how many events are ready for the given rule.
  int countReadyEvents(Rule rule) {
    var count = 0;
    for (var i = 0; i < queue.length; i++) {
      if (queue[i].pendingConsumers.contains(rule)) {
        count++;
      }
    }
    return count;
  }

  /// Peek at the first event ready for the given rule (without consuming).
  PulseEvent<T>? peek(Rule rule) {
    for (var i = 0; i < queue.length; i++) {
      if (queue[i].pendingConsumers.contains(rule)) {
        return queue[i];
      }
    }
    return null;
  }

  /// Consume the first event ready for the given rule.
  /// Marks the event as consumed by this rule.
  /// If all consumers have processed the event, it is removed and recycled.
  PulseEvent<T> consume(Rule rule) {
    for (var i = 0; i < queue.length; i++) {
      final ev = queue[i];
      if (ev.pendingConsumers.contains(rule)) {
        ev.pendingConsumers.remove(rule);
        // GC: if no more pending consumers, remove from queue.
        if (ev.pendingConsumers.isEmpty) {
          queue.removeAt(i);
        }
        return ev;
      }
    }
    throw StateError('Mailbox(${type.name}): no ready event for rule ${rule.id}');
  }

  /// Reverse a consumption -- put events back as pending for this rule.
  /// Used when a guard fails after consumption.
  void unconsumeAll(Rule rule, List<PulseEvent> events) {
    for (final ev in events) {
      ev.pendingConsumers.add(rule);
      if (!queue.contains(ev)) {
        queue.add(ev as PulseEvent<T>);
      }
    }
  }

  /// Get the number of events in the queue.
  int get size => queue.length;

  /// Clear all events.
  void clear() {
    for (final ev in queue) {
      recycleEvent(ev);
    }
    queue.clear();
  }
}

File: packages/core/src/mailbox.ts
import type { EventType, PulseEvent, Rule } from './types.js'
import { recycleEvent } from './event.js'

/**
 * Mailbox — holds pending events for a given EventType.
 * Events remain in the mailbox until all registered consumers have processed them.
 * Stale events (older than TTL) are evicted to prevent unbounded growth from incomplete joins.
 */
export class Mailbox<T = any> {
  type: EventType<T>
  queue: PulseEvent<T>[] = []

  constructor(type: EventType<T>) {
    this.type = type
  }

  /** Add an event to this mailbox */
  enqueue(event: PulseEvent<T>): void {
    this.queue.push(event)
  }

  /** Evict orphaned events whose pending consumers have all been disposed */
  evictOrphans(): void {
    let i = 0
    while (i < this.queue.length) {
      const ev = this.queue[i]
      // Remove disposed consumers
      for (const rule of ev._pendingConsumers) {
        if (rule._disposed) {
          ev._pendingConsumers.delete(rule)
        }
      }
      // If no consumers remain, event is orphaned
      if (ev._pendingConsumers.size === 0) {
        this.queue.splice(i, 1)
        recycleEvent(ev)
      } else {
        i++
      }
    }
  }

  /** Check if there is at least one event ready for the given rule */
  hasReadyEvent(rule: Rule): boolean {
    for (let i = 0; i < this.queue.length; i++) {
      if (this.queue[i]._pendingConsumers.has(rule)) {
        return true
      }
    }
    return false
  }

  /** Count how many events are ready for the given rule */
  countReadyEvents(rule: Rule): number {
    let count = 0
    for (let i = 0; i < this.queue.length; i++) {
      if (this.queue[i]._pendingConsumers.has(rule)) {
        count++
      }
    }
    return count
  }

  /** Peek at the first event ready for the given rule (without consuming) */
  peek(rule: Rule): PulseEvent<T> | null {
    for (let i = 0; i < this.queue.length; i++) {
      if (this.queue[i]._pendingConsumers.has(rule)) {
        return this.queue[i]
      }
    }
    return null
  }

  /**
   * Consume the first event ready for the given rule.
   * Marks the event as consumed by this rule.
   * If all consumers have processed the event, it is removed and recycled.
   */
  consume(rule: Rule): PulseEvent<T> {
    for (let i = 0; i < this.queue.length; i++) {
      const ev = this.queue[i]
      if (ev._pendingConsumers.has(rule)) {
        ev._pendingConsumers.delete(rule)
        // GC: if no more pending consumers, remove from queue.
        // Do NOT recycle here — the caller still needs the payload.
        if (ev._pendingConsumers.size === 0) {
          this.queue.splice(i, 1)
        }
        return ev
      }
    }
    throw new Error(`Mailbox(${this.type.name}): no ready event for rule ${rule.id}`)
  }

  /**
   * Reverse a consumption — put events back as pending for this rule.
   * Used when a guard fails after consumption.
   */
  unconsumeAll(rule: Rule, events: PulseEvent[]): void {
    for (const ev of events) {
      ev._pendingConsumers.add(rule)
      // If the event was already removed from the queue (recycled), re-add it.
      // This happens when the consuming rule was the last consumer.
      if (!this.queue.includes(ev)) {
        this.queue.push(ev)
      }
    }
  }

  /** Get the number of events in the queue */
  get size(): number {
    return this.queue.length
  }

  /** Clear all events */
  clear(): void {
    for (const ev of this.queue) {
      recycleEvent(ev)
    }
    this.queue.length = 0
  }
}
